File: projects/15-data-center-environmental-monitoring/fog/aggregation.js
"use strict";

// Pure window-aggregate math over a ring buffer's ordered-array snapshot.
// Kept separate from ringBuffer.js so it can be unit tested against plain
// arrays of {ts, value} readings without touching the ring's write-index
// bookkeeping at all.
function summarizeWindow(sensorType, siteId, unit, readings, windowStart, windowEnd) {
  const values = readings.map((r) => r.value);
  const sum = values.reduce((acc, v) => acc + v, 0);
  return {
    sensor_type: sensorType,
    site_id: siteId,
    unit,
    window_start: windowStart,
    window_end: windowEnd,
    count: values.length,
    min: Math.min(...values),
    max: Math.max(...values),
    avg: Math.round((sum / values.length) * 1000) / 1000,
    // last-in-order, not max-timestamp -- ringToOrderedArray already hands
    // this function readings in original write order, so the last array
    // element is genuinely the most recently sampled value.
    latest: readings[readings.length - 1].value,
  };
}

module.exports = { summarizeWindow };

File: projects/14-smart-parking-management/fog/aggregation.py
from statistics import fmean


def aggregate(sensor_type, site_id, unit, readings, window_start, window_end):
    """Reduce one window's raw readings into a single summary; latest is the last reading in arrival order."""
    values = [r["value"] for r in readings]
    return {
        "sensor_type": sensor_type,
        "site_id": site_id,
        "unit": unit,
        "window_start": window_start,
        "window_end": window_end,
        "count": len(values),
        "min": min(values),
        "max": max(values),
        "avg": round(fmean(values), 3),
        "latest": readings[-1]["value"],
    }
